### ISSUE TITLE
The condition variables in APIv2 still have spurious wakeups when there are many threads

### ISSUE THREAD

**Issue by hxzrx** [sionescu/bordeaux-threads#109]
Recently I intensively tested my queue lib with condition variables,  but I found that it still failed even with APIv2.
Relate ISSUES #29 

I wrote a test .
The test was rather simple:
1. The concurrent-push function spawns threads to push integers into a list and notify a condition variable.
2. The concurrent-pop function spawns threads to pop from that list, but when it finds a nil value was popped, push nil to another list atomically.
3. In the end, inspect the latter list and check if nil value did been pushed.

In my laptop, when the count of threads was 1000, the test seldom failed, but when the count kept increasing, the failing rate increased accordingly.

According to ISSUES #29 , another function concurrent-pop2 which uses a loop to wait cvar, and the tests passed (up to 50k threads), although it's slow for other reasons.

My test env:
openSUSE Leap 15.3 hosted on win10, the vm has 4 cores and 16G memory.
SBCL 2.2.1.

The codes are pasted below:
```
(defparameter *lst* (list))
(defparameter *fail-results* (list))
(defparameter *lock* (bt2:make-lock))
(defparameter *cvar* (bt2:make-condition-variable))

(defun concurrent-push (n)
  "Push into *lst* and notify *cvar*."
  (dotimes (i n)
    (bt2:make-thread #'(lambda ()
                         (bt2:with-lock-held (*lock*)
                           (push (random 10) *lst*)
                           (bt2:condition-notify *cvar*))))))

(defun concurrent-pop (n)
  "Pop if *lst* is not null,                                                                                                         
else wait condition *cvar*,                                                                                                          
but if a nil result was popped, push this result into *fail-results* atomically."
  (dotimes (i n)
    (bt2:make-thread #'(lambda ()
                         (bt2:with-lock-held (*lock*)
                           (if *lst*
                               (pop *lst*)
                               (progn (bt2:condition-wait *cvar* *lock*)
                                      (let ((res (pop *lst*)))
                                        (unless res
                                          (sb-ext:atomic-push res *fail-results*))
                                        res))))))))

(defun concurrent-pop2 (n)
  "Like concurrent-pop, but use a loop when waiting cvar."
  (dotimes (i n)
    (bt2:make-thread #'(lambda ()
                         (bt2:with-lock-held (*lock*)
                           (if *lst*
                               (pop *lst*)
                               (progn
                                 (loop while (null *lst*)
                                       do (bt2:condition-wait *cvar* *lock*))
                                 (let ((res (pop *lst*)))
                                   (unless res
                                     (sb-ext:atomic-push res *fail-results*))
                                   res))))))))

(defun test (n)
  "test with concurrent-pop which has no loop"
  (time
   (progn
     (format t "Begin testing cvar!~%")
     (setf *lst* (list))
     (setf *fail-results* (list))
     (concurrent-pop n)
     (concurrent-push n)
     (if *fail-results*
         (format t "Test failed: ~d~%" *fail-results*)
         (format t "Test passed!~%~%")))))

(defun test2 (n)
  "test with concurrent-pop2 which has a loop to wait cvar"
  (time
   (progn
     (format t "Begin testing cvar!~%")
     (setf *lst* (list))
     (setf *fail-results* (list))
     (concurrent-pop2 n)
     (concurrent-push n)
     (if *fail-results*
         (format t "Test failed: ~d~%" *fail-results*)
         (format t "Test passed!~%~%")))))

(test 5000)
```

**Comment by sionescu** [comment]
My dev laptop is a Leap 15.3 with a 4-core Intel E3-1505M v5 and all tests pass even with 10k threads. That said, spurious wakeups are part of the API of POSIX condition variables.

**Comment by hxzrx** [comment]
> My dev laptop is a Leap 15.3 with a 4-core Intel E3-1505M v5 and all tests pass even with 10k threads. That said, spurious wakeups are part of the API of POSIX condition variables. 

I tested it again with the latest sbcl 2.2.3, and it still failed. 
I think it's better to wrap condition-wait into a loop, as  the inevitable of missed signals or spurious wakeups, and the SBCL's documentation also gives a clear indication of this.

**Comment by sionescu** [comment]
Yes, calling `condition-wait` in a loop is the correct way ot use it. I'm not sure what this issue is about.

**Comment by hxzrx** [comment]
> Yes, calling `condition-wait` in a loop is the correct way ot use it. I'm not sure what this issue is about.

This issue is a sequel of issue #29 , where  Bergmann gave an popq function which used condition-wait without a loop, and afterwards, he replied that apiv2 worked. 
I went into Bergmann's cl-gserver [code](https://github.com/mdbergmann/cl-gserver/blob/master/src/queue/queue.lisp) and found the popq function there, and I also found there's once a loop that wrapped condition-wait, but the whole loop was commented. 
And thus I mistook api2's condition-wait could be utilized without a loop, that's all about the raising of issue. 
Oh, Leap 15.3 too! Cheers!
